Import weighted stats and moments from StatsBase to Statistics

nalimilan opened the pull request:
This includes methods for `mean`, `quantile`, `median`, `var`, `std`, `cov` and `cor`, plus new functions `skewness` and `kurtosis`, and weight types. Code is copied from StatsBase with some cleanup where needed, in particular for dispatch, to move from `@nloops`/`@nrefs` to cartesian indexing and to
be closer to the `mapreducedim` code. Weights are now passed via a keyword argument rather than by dispatching on `AbstractWeights`, so as to support any array where all weights types give the same result.

Still to address:
- [x] Do we want `mean_and_var` and `mean_and_std`? It's not too hard to do `m = mean(...); s = std(..., mean=m)` manually. Also the names aren't great. UPDATE: dropped.
- [x] Do we want `moment`? It feels redundant with `mean`, `var`, `skewness` and `kurtosis`. UPDATE: dropped.
- [x] Should we also support weighted `sum`? For now I've kept `wsum` internal (it is used by `mean`), it can always be implemented later in Base. UPDATE: added.
- [x] Check that performance didn't regress when porting from `@nloop`/`@nref` to cartesian indexing. UPDATE: performance is similar, and even sometimes faster.
- [x] Clean tests (notably ensuring that non-vector and non-`AbstractWeights` weights are supported).
- [x] Check that everything is clean, in particular docstrings.

Closes https://github.com/JuliaLang/julia/issues/29974. See https://github.com/JuliaLang/julia/pull/27152#issuecomment-390687566 for an outline of a broader roadmap.

nalimilan committed:
Import weighted stats and moments from StatsBase to Statistics

This includes methods for mean, quantile, median, var, std, cov and cor,
plus new functions skewness and kurtosis, and weight types.
Code is copied from StatsBase with some cleanup where needed, in particular
for dispatch, to move from `@nloops`/`@nrefs` to cartesian indexing and to
be closer to the mapreducedim code.
Weights are now passed via a keyword argument rather than by dispatching on
AbstractWeights, so as to support any array where all weights types
give the same result.

ararslan commented:
> Do we want `mean_and_var` and `mean_and_std`? It's not too hard to do `m = mean(...); s = std(..., mean=m)` manually. Also the names aren't great.

👍 to dumping them.

> Do we want `moment`? It feels redundant with `mean`, `var`, `skewness` and `kurtosis`.

Unless we're going to support arbitrary moments (or just 5th or higher), I'd say leave it out.

> Should we also support weighted `sum`?

👍 to supporting it.

rofinn commented:
Out of curiosity, what's the advantage with having this in Statistics? I only ask because there are still some open issues with the weights type in StatsBase that we might want to address (e.g., n-dimensional weights, mutability of weights).

ararslan commented:
> Out of curiosity, what's the advantage with having this in Statistics?

One big one is better APIs, for example `mean(x, weights=w)`, which also ensures that cases like https://github.com/JuliaStats/StatsBase.jl/issues/475 don't occur.

rofinn commented:
Oh, yeah, I guess that makes sense. Looks like github also supports transferring issues to another repo https://help.github.com/en/articles/transferring-an-issue-to-another-repository. I would prefer that we try to retain the history for the files that we're actively copying over.

nalimilan commented:
Another advantage is to eventually remove StatsBase, whose name is confusing since it's not more "base" than Statistics (quite the contrary actually).

>  I would prefer that we try to retain the history for the files that we're actively copying over.

I'm not sure that's really possible, as the code cannot be imported as-is and pass tests. So we would have to keep broken commits in history, which isn't great for bisecting. We can still refer to the StatsBase repo for history, though.

oxinabox commented on stdlib/Statistics/src/moments.jl:
```diff
@@ -0,0 +1,297 @@
+    s / sum(w)
+end
+
+function _momentk(A::AbstractArray, k::Int, m::Real)
```
These are not nice,
can't we do something better than this?
even if it is `@generated _moments(::Val{K}, A::AbstractArray, w::AbstractArray, m::Real) where K`

nalimilan replied:
I guess so, but let's decide whether we want to keep this function first (see OP). Is it really common to compute moments beyond order 4 (kurtosis)?

oxinabox replied:
Sorry I was unclear.
I was suggesting replacing monents 1-4 with something that wasn't hand coded.
(Though if you didn't want to go beyond 4, can use `@eval` I guess)

nalimilan committed:
Remove moment and combined stats, make other functions accept any iterable

nalimilan committed:
Implement weighted sum

nalimilan committed:
Move optimized weighted sum methods to LinearAlgebra

nalimilan committed:
More tests

nalimilan committed:
Move varcorrection to Statistics.jl

nalimilan committed:
Docs

nalimilan committed:
Performance fix

nalimilan committed:
Cleanup

nalimilan committed:
Fix bug

nalimilan committed:
Fix TODO

nalimilan committed:
Another cleanup, enable more tests

nalimilan commented:
I've implemented `sum(x; weights, dims)` in Base. This is tricky since StatsBase had optimized methods using BLAS, which make a large difference for performance. The solution I found is to have the generic fallbacks in reducedim.jl, and add methods for `BlasReal` from LinearAlgebra.

I've also cleaned a few things and checked that performance hasn't regressed, so I think the PR is now ready for a serious review. This is a large piece of code, and the reductions are particularly tricky, so double-checking would really be appreciated.

Tokazama commented on stdlib/Statistics/docs/src/index.md:
```diff
@@ -4,22 +4,81 @@
-The Statistics module contains basic statistics functionality.
+The Statistics module contains basic statistics functionality: mean, median, quantiles,
+standard deviation, variance, skewness, kurtosis, correlation and covariance.
+Statistics can be weighted, and several weights types are distinguished to apply appropriate
```
Perhaps this should read "several weight types" instead of "several weights types". If you're referring to the actual type then " several `AbstractWeights` types" maybe?

nalimilan replied:
Why "weight type"? The plural sounds more appropriate since weights only make sense as a series of values. Though "types of weights" might be better.

Tokazama replied:
It just sounded weird to me. I think "types of weights" is what sounds best.

rofinn commented on stdlib/Statistics/src/Statistics.jl:
```diff
@@ -294,22 +350,111 @@ over dimensions, and `m` may contain means for each dimension of `itr`.
 
 """
-    var(itr; dims, corrected::Bool=true, mean=nothing)
+    var(itr; corrected::Bool=true, [weights::AbstractWeights], [mean], [dims])
```
Maybe we should support `weights::AbstractArray` for consistency with other methods like `mean`? We could always convert non-weight arrays with `Weights(weights)`.

nalimilan replied:
Actually we can't support arbitrary vectors since there's no `varcorrection` for them. `Weights` isn't supported.

rofinn replied:
That's true, but you should still be able to do `var(itr; weights=..., corrected=false)`.

rofinn commented on stdlib/Statistics/src/Statistics.jl:
```diff
@@ -294,22 +350,111 @@ over dimensions, and `m` may contain means for each dimension of `itr`.
+    varcorrection(w::Weights, corrected=false)
+
+Returns ``\\frac{1}{\\sum w}`` when `corrected=false` and throws an `ArgumentError`
+if `corrected=true`.
```
Might be good to include links here as well (e.g., https://en.wikipedia.org/wiki/Weighted_arithmetic_mean)

rofinn commented on stdlib/Statistics/src/Statistics.jl:
```diff
@@ -294,22 +350,111 @@ over dimensions, and `m` may contain means for each dimension of `itr`.
+    varcorrection(w::AnalyticWeights, corrected=false)
+
+* `corrected=true`: ``\\frac{1}{\\sum w - \\sum {w^2} / \\sum w}``
+* `corrected=false`: ``\\frac{1}{\\sum w}``
```
Might be good to include links here as well (e.g., https://en.wikipedia.org/wiki/Weighted_arithmetic_mean)

rofinn commented on stdlib/Statistics/src/Statistics.jl:
```diff
@@ -963,11 +1186,87 @@ julia> quantile(skipmissing([1, 10, missing]), 0.5)
+
+    wsum = sum(w)
+    wsum == 0 && throw(ArgumentError("weight vector cannot sum to zero"))
+    length(v) == length(w) || throw(ArgumentError("data and weight vectors must be the same size," *
```
If you need to split it over multiple lines then you might want to just use an `if`?

rofinn commented on stdlib/Statistics/src/Statistics.jl:
```diff
@@ -963,11 +1186,87 @@ julia> quantile(skipmissing([1, 10, missing]), 0.5)
+        x < 0 && throw(ArgumentError("weight vector cannot contain negative entries"))
+    end
+
+    isa(w, FrequencyWeights) && !(eltype(w) <: Integer) && any(!isinteger, w) &&
```
Again, maybe just use an `if` block.

rofinn commented on stdlib/Statistics/src/weights.jl:
```diff
@@ -0,0 +1,170 @@
+end
+
+Base.isequal(x::AbstractWeights, y::AbstractWeights) = false
+Base.:(==)(x::AbstractWeights, y::AbstractWeights)   = false
```
Missing a newline at end of file.

rofinn reviewed (changes requested):
> I'm not sure that's really possible, as the code cannot be imported as-is and pass tests. So we would have to keep broken commits in history, which isn't great for bisecting. We can still refer to the StatsBase repo for history, though.

How do you propose we refer to the StatsBase history? FWIW, I'd prefer that we retain history for this data type as a lot of changes/decisions were made while they were being developed. It is also helpful for github's suggested reviewers feature. Finally, I think this would be better to add after the stdlibs have been moved into separate packages as julia can just pin the Statistics version to a particular release, but folks can easily free/update to these new features regardless of which julia version they're using. It'll also be easier add features and bug fixes for these types without depending on the next julia release.

nalimilan commented:
> How do you propose we refer to the StatsBase history? FWIW, I'd prefer that we retain history for this data type as a lot of changes/decisions were made while they were being developed. It is also helpful for github's suggested reviewers feature.

As I said, I don't know whether that's possible, and I'm not aware of a precedent. Do you have ideas?

>  Finally, I think this would be better to add after the stdlibs have been moved into separate packages as julia can just pin the Statistics version to a particular release, but folks can easily free/update to these new features regardless of which julia version they're using. It'll also be easier add features and bug fixes for these types without depending on the next julia release.

What would be the advantage of waiting? We can keep exporting the types from StatsBase (and just re-exporting when Statistics provides them), so that it keeps working as it does on all Julia versions. Then once stdlibs can be versioned, people will be able to start depending only on a given Statistics version, and drop the StatsBase dependency. But waiting blocks any progress on the StatsBase front (this PR is just a small part of the needed work).

Tokazama commented on stdlib/Statistics/src/moments.jl:
```diff
@@ -0,0 +1,151 @@
+    cm3 = cm2 * z # empirical 3rd centered moment
+    n = 1
+    y = iterate(x, s)
+    while y !== nothing
```
Is the reason that you can't use `@inbounds @simd for i in eachindex(A)` because you're starting on the 2nd iterate in this loop?

rofinn commented:
> As I said, I don't know whether that's possible, and I'm not aware of a precedent. Do you have ideas?

Yes, we do this a reasonable amount at work.

1. You'd use `git filter-branch` to isolate the files you want to merge into Statistics locally (don't push this).
2. Add your local stats base repo as a remote to the local julia repo
3. Use `git pull --allow-unrelated-histories` from you branch in the julia repo to pull in those files from the StatsBase repo. Then you'd just need to move the file to where you'd like and apply the changes for the new API that you like.

NOTE: The reason I'd prefer to wait till Statistics is a separate repo is that it'll make the history cleaner because the `src/weights.jl` file will be pulled into the right location by default. It should also set a nice precedent for moving features from StatsBase.jl to Statistics.jl.

nalimilan commented:
Let's continue this at https://github.com/JuliaStats/Statistics.jl/pull/2.

iamed2 commented:
> The reason I'd prefer to wait till Statistics is a separate repo is that it'll make the history cleaner because the src/weights.jl file will be pulled into the right location by default. 

You can pull the history in and have it match up by using `git subtree` with a prefix. I think moving it out is still better, but I just want everyone to know that it's possible to do this history merge without moving it out.

pdeffebach commented:
Can someone give me an update on what work needs to be done on this? Is it all waiting on https://github.com/JuliaLang/Statistics.jl/pull/2? 

Or are there design decisions that need to be made still?

nalimilan commented:
Continuation is at JuliaLang/Statistics.jl#2.
